File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
const isValid = (s) => {
    let stack = [];
    const bracketMap = {')': '(', ']': '[', '}': '{'};
    
    for(const ch of s){
        if('([{'.includes(ch)) stack.push(ch)
        else {
            if(stack.length === 0 || stack[stack.length - 1] !== bracketMap[ch]) return false;
            stack.pop();
        }
    }
    return stack.length === 0;
}

function solution(s) {
    let cnt = 0;
    for(let i = 0 ; i < s.length ; i++){
        const rotated = s.slice(i) + s.slice(0, i);
        if(isValid(rotated)) cnt++;
    }
    return cnt;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def is_valid(s):
    stack = []
    bracket_map = {')': '(', ']': '[', '}': '{'}

    for ch in s:
        if ch in '([{':
            stack.append(ch)
        else:
            if not stack or stack[-1] != bracket_map[ch]:
                return False
            stack.pop()
    return not stack

def solution(s):
    count = 0
    n = len(s)

    for i in range(n):
        rotated = s[i:] + s[:i]
        if is_valid(rotated):
            count += 1

    return count